lib/permission.js

```diff
@@ -17,21 +23,159 @@ module.exports = api;
 
 var logger = bedrock.loggers.get('app');
 
-bedrock.events.on('bedrock.start', init);
+bedrock.events.on('bedrock-mongodb.ready', function(callback) {
+  PERMISSIONS = bedrock.config.permission.permissions;
+  async.auto({
+    openCollections: function(callback) {
+      database.openCollections(['roles'], callback);
+    },
+    createIndexes: ['openCollections', function(callback) {
+      database.createIndexes([{
+        collection: 'roles',
+        fields: {id: 1},
+        options: {unique: true, background: false}
+      }], callback);
+    }],
+    addRoles: ['createIndexes', function(callback) {
+      async.eachSeries(
+        bedrock.config.permission.roles, function(role, callback) {
+        role.permanent = true;
+        logger.debug('adding role', role);
+        api.addRole(null, role, function(err) {
+          if(err && database.isDuplicateError(err)) {
+            err = null;
+          }
+          callback(err);
+        });
+      }, callback);
+    }]
+  }, callback);
+});
 
-function init(callback) {
-  // log permissions
-  var permissions = bedrock.config.permission.permissions;
-  Object.keys(permissions).forEach(function(permission) {
-    logger.debug('adding permission', permissions[permission]);
-  });
-  // log roles
-  var roles = bedrock.config.permission.roles;
-  Object.keys(roles).forEach(function(role) {
-    logger.debug('adding role', roles[role]);
-  });
-  callback();
-}
+/**
+ * Returns a role based on the id.
+ *
+ * @param id The id of the role.
+ * @param callback(err, role) Called once the operation completes.
+ */
+api.getRole = function(actor, id, callback) {
+  async.waterfall([
+    function(callback) {
+      api.checkPermission(actor, PERMISSIONS.IDENTITY_ADMIN, callback);
```

Comment:
bedrock-permission has no dependency on bedrock-identity, which is what defines `PERMISSIONS.IDENTITY_ADMIN` and can't have one, it would cause a circular dependency. This permission needs to be changed, probably to a new permission, `ROLE_ACCESS`.